runtime/src/main/java/org/corfudb/runtime/clients/NettyClientProtobufRouter.java

```
@@ -0,0 +1,360 @@
+        try {
+            // Mark this message as Protobuf message (temporarily)
+            requestOutputStream.writeByte(API.PROTO_CORFU_MSG_MARK);
+            request.writeTo(requestOutputStream);
+            ctx.writeAndFlush(outBuf, ctx.voidPromise());
+            log.trace("Sent one-way request message: {}", request.getHeader());
+        } catch (IOException e) {
+            log.warn("sendRequest[{}]: Exception occurred when sending request {}, caused by {}",
+                    request.getHeader().getRequestId(), request.getHeader(), e.getCause(), e);
+        } finally {
+            IOUtils.closeQuietly(requestOutputStream);
+        }
+    }
+
+    /**
+     * Validate the clientID of a CorfuMsg.
+     *
+     * @param header The header of incoming message used for validation.
+     * @return True, if the clientID is correct, but false otherwise.
+     */
+    private boolean validateClientId(UUID clientId) {
+        // Check if the message is intended for us. If not, drop the message.
+        if (!clientId.equals(parameters.getClientId())) {
+            log.warn("Incoming message intended for client {}, our id is {}, dropping!",
+                    clientId, parameters.getClientId());
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Read data from the Channel.
+     *
+     * @param ctx Channel handler context
+     * @param msg Object received in inbound buffer
+     * @throws Exception Exception thrown during channelRead
+     */
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ByteBuf msgBuf = (ByteBuf) msg;
+
+        // Temporary -- If message is a legacy message, forward the message.
+        byte msgMark = msgBuf.getByte(msgBuf.readerIndex());
+        if (msgMark == API.LEGACY_CORFU_MSG_MARK) {
+            // TODO should we also use this method on the client side?
+            ctx.fireChannelRead(msgBuf);
+            return;
```

This is not needed here. It was needed server-side since both types of routers are in my pipeline. On the client-side, you won't have an older router sitting in the pipeline, so there is no need to forward the message. If the new client receives a legacy message as a Response, we can throw an exception.

---

Got you, thanks!